stablehlo/transforms/StablehloLegalizeQuantToInt.cpp

```diff
@@ -0,0 +1,1385 @@
+/* Copyright 2024 The StableHLO Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <cstdint>
+#include <cstdlib>
+#include <memory>
+#include <optional>
+#include <utility>
+#include <variant>
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Func/Transforms/FuncConversions.h"
+#include "mlir/Dialect/Quant/QuantOps.h"
+#include "mlir/Dialect/Quant/QuantTypes.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypeInterfaces.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Location.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/OpDefinition.h"
+#include "mlir/IR/OperationSupport.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Region.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "stablehlo/dialect/ChloOps.h"
+#include "stablehlo/dialect/StablehloOps.h"
+
+namespace mlir::stablehlo {
+namespace {
+
+// TODO: b/311218165 - consider extract this to common utils and better ways to
+// handle polymorphism.
+using QuantType = std::variant<quant::UniformQuantizedType,
```

Comment:
Note for future refactoring, I don't love the variant usage. In MLIR we tend to prefer verification / getters with failure, etc. Let's keep it for now though.

Reply:
I.e.

```
bool isPerChannelType(QuantType quantType) {
  return std::holds_alternative<quant::UniformQuantizedPerAxisType>(quantType);
}
```

MLIR type has facilities built in to make `isa<...>(type)` extremely fast.

Reply:
Absolutely.  The variant is used as a catch all for both per-tensor and per-axis types. We have the base class `quant::QuantizedType` which would serve the same purpose. I believe the pass can be cleaned up a lot w.r.t the how the Q types are handled. Let us have a follow up PR to address that.

Reply:
Here is the follow up: https://github.com/openxla/stablehlo/pull/2385